alibi_detect/od/pytorch/svm.py

```diff
@@ -0,0 +1,307 @@
+            - lower_bound: loss lower bound.
+        """
+
+        self.nystroem = _Nystroem(
+            self.kernel,  # type: ignore[arg-type]
+            self.n_components
+        )
+
+        X_nys = self.nystroem.fit(x_ref).transform(x_ref)
+        n, d = X_nys.shape
+        min_eta, max_eta = step_size_range
+        etas = torch.tensor(
+            np.linspace(
+                np.log(min_eta),
+                np.log(max_eta),
+                n_step_sizes
+            ),
+            dtype=X_nys.dtype,
+            device=self.device
+        ).exp()
+
+        # Initialise coeffs/preds/loss
+        coeffs = torch.zeros(d, dtype=X_nys.dtype, device=self.device)
+        intercept = torch.zeros(1, dtype=X_nys.dtype, device=self.device)
+        preds = X_nys @ coeffs + intercept
+        loss = nu * (coeffs.square().sum()/2 + intercept) + hinge_loss(preds)
+        min_loss, min_loss_coeffs, min_loss_intercept = loss, coeffs, intercept
+        iter, t_since_improv = 0, 0
+        converged = False
+
+        with tqdm(total=max_iter, disable=not verbose) as pbar:
+            while not converged:
+                pbar.update(1)
+                # First two lines give form of sgd update (for each candidate step size)
+                sup_vec_inds = (preds < 1)
+                cand_coeffs = coeffs[:, None] * \
+                    (1-etas*nu) + etas*(X_nys[sup_vec_inds].sum(0)/n)[:, None]
+                cand_intercept = intercept - etas*nu + (sup_vec_inds.sum()/n)
+
+                # Compute loss for each candidate step size and choose the best
+                cand_preds = X_nys @ cand_coeffs + cand_intercept
+                cand_losses = nu * (cand_coeffs.square().sum(0)/2 + cand_intercept) + hinge_loss(cand_preds)
+                best_step_size = cand_losses.argmin()
+                coeffs, intercept = cand_coeffs[:, best_step_size], cand_intercept[best_step_size]
+                preds, loss = cand_preds[:, best_step_size], cand_losses[best_step_size]
+
+                # Keep track of best performing coefficients and time since improving (by more than tol)
+                if loss < min_loss:
+                    if loss < min_loss - tol:
+                        t_since_improv = 0
+                    min_loss, min_loss_coeffs, min_loss_intercept = loss, coeffs, intercept
+                else:
+                    t_since_improv += 1
+
+                # Decide whether to continue
+                if iter > max_iter or t_since_improv > n_iter_no_change:
+                    self.coeffs = min_loss_coeffs
+                    self.intercept = min_loss_intercept
+                    converged = True
+                    break
+                else:
+                    iter += 1
+
+                if verbose and isinstance(pbar, tqdm):
+                    pbar.set_postfix(dict(loss=loss.cpu().detach().numpy().item()))
+
+        self._set_fitted()
+        return {
```

Comment:
Not against returning a dict like this for the `fit` method but if so we should make sure we're consistent across detectors (e.g. currently the GMM detector doesn't do this)